system/lib/wasmfs/wasmfs.cpp

```diff
@@ -134,7 +140,9 @@ void WasmFS::preloadFiles() {
 
     parentDir->locked().setEntry(base, created);
 
-    created->locked().preloadFromJS(i);
+    // Need to cast to a MemoryFile currently since preloadFromJS is defined
+    // there and rootBackend->createFile returns a DataFile.
+    created->dynCast<MemoryFile>()->locked().preloadFromJS(i);
```

Comment:
It doesn't seem great to mix trying to be general by using `rootBackend`, but then needing to downcast to `MemoryFile` anyway. It seems harder to make mistakes in the future if we explicitly use `MemoryFile` or at least `MemoryBackend` the entire time.

Also, this `dynCast` checks that the file is some subclass of `DataFile`, but doesn't specifically check that it is a MemoryFile because `MemoryFile` does not have its own `kind`. So this downcast also seems somewhat dangerous.

Reply:
I see, yeah the downcast here is not safe. I will switch it back to a hardcoded `MemoryFile` as we discussed with the TODO comment.

Reply:
We decided to generalize this in a follow up PR.